File: MATLAB/Pulse_Reconstruction.m
%% Rectangular Pulse Function Overlap presentation + Gibbs Phenomena

%Parameters declaration:
T1=1;T=4;N=[1,2,3,4,5,6];

%Function definition:
time = -T/2:0.01:T/2;    
    
x = zeros(size(time));
x( (time > -T1) & (time < T1) ) = 1;

%The color vector below is critical for an efficient data visualization
Colors = ['y','m','c','r','g','b'];
    
figure;
plot(time,x,'k');title('Pulse Function Reconstruction');hold on;
legend('Original Signal')
%The following loop generates the legend for each function accordingly
for j=1:length(N)
   fReconsta = FindFourierCoeff(time, x, N(j));
   txt=['Reconstructed Signal N=',num2str(N(j))];
   plot(time, fReconsta,Colors(j),'DisplayName',txt);
   hold on;         
end
hold off;
legend show;
function [fReconst] = FindFourierCoeff(time, fOrig, k)
    dx = time(2) - time(1);
    a0 = (1/pi) .* sum(fOrig.*dx);
        
    sumVal = 0;
    aN = zeros(k, 1);
    bN = zeros(k, 1);
    for iter = 1:k
        aN(iter) = (1/pi) .* sum(fOrig.*cos(iter*time).*dx);
        bN(iter) = (1/pi) .* sum(fOrig.*sin(iter*time).*dx);
        
        sumVal = sumVal + aN(iter).*cos(iter.*time) + bN(iter).*sin(iter.*time);
    end
    
    fReconst = a0/2 + sumVal;

end

File: MATLAB/Pulse_Reconstruction.py
import numpy as np
import matplotlib.pylab as plt


T1=1;T=4;N=[1,3,7,19,79]

#Function definition:
time = np.arange(-T/2,T/2, 0.01) 
    
x = np.zeros(len(time))
x[np.where( (time > -T1) & (time < T1) )] = 1

#The color vector below is critical for an efficient data visualization
Colors = ['y','m','c','r','g','b']
    
plt.figure()
plt.plot(time,x,'k', label='Original Signal');plt.title('Pulse Function Reconstruction')

def FindFourierCoeff(time, fOrig, k):
    dx = time[2] - time[1]
    a0 = (1/np.pi) * sum(fOrig*dx)
        
    sumVal = 0
    aN = np.zeros(k)
    bN = np.zeros(k)
    for iter in range(k):
        aN[iter] = (1/np.pi) * sum(fOrig*np.cos((iter+1)*time)*dx)
        bN[iter] = (1/np.pi) * sum(fOrig*np.sin((iter+1)*time)*dx)
        
        sumVal = sumVal + aN[iter]*np.cos((iter+1)*time) + bN[iter]*np.sin((iter+1)*time)
        
    fReconst = a0/2 + sumVal
    return fReconst

for j in range(len(N)):
    fReconsta = np.zeros(N[j])
    fReconsta = FindFourierCoeff(time, x, N[j])
    txt=['Reconstructed Signal N='+str(N[j])]
    plt.plot(time, fReconsta, color=Colors[j],label='txt')
    #plt.legend(txt)

plt.show()   
